### PR TITLE
OSDOCS-5387: add kubeconfig management

### PR DESCRIPTION
**Pull request by ShaunaDiaz**
Version(s):
4.13+

Issue:
https://issues.redhat.com/browse/OSDOCS-5387

Link to docs preview:
https://57834--docspreview.netlify.app/microshift/latest/microshift_configuring/microshift-kubeconfig.html

QE review:
- [x] QE has approved this change.

### CODE REVIEW

File: modules/microshift-kubeconfig-local-access.adoc
Comment (dhellmann) [inline modules/microshift-kubeconfig-local-access.adoc:9]:
```suggestion
The local access `kubeconfig` file is generated in the `/var/lib/microshift/resources/kubeadmin` directory. This configuration file allows access to the local API server using `localhost` and the internal certificate authority (CA) of `localhost`. The CA of the local access `kubeconfig` file cannot be used outside of the {product-title} host.
```

@pacevedom is that right, the local kubeconfig file can't be copied to another host? If so, we should be more explicit about that in the statement here for someone who does not know what a "CA" is.

File: modules/microshift-kubeconfig-local-access.adoc
Comment (dhellmann) [inline modules/microshift-kubeconfig-local-access.adoc:11]: This paragraph starts out with "the cluster" as the actor and then gives the user instructions ("you must"). Maybe we should split it up to explain how the cluster does authentication internally, then tell the user to use a specific kubeconfig when trying to access the API server from the same host.

File: modules/microshift-kubeconfig-local-access.adoc
Comment (dhellmann) [inline modules/microshift-kubeconfig-local-access.adoc:13]: I don't understand what the first sentence here means. What values are taking from `config.yaml`?

Maybe instead of focusing on what users can and cannot do with specific files, this would be clearer if we showed them how to do things we think they are going to want to do?

* Access the API server from the host running MicroShift
* Access the API server from another host using the IP of the MicroShift host
* Access the API server from another host using the hostname of the MicroShift host

What else?

File: modules/microshift-kubeconfig-overview.adoc
Comment (dhellmann) [inline modules/microshift-kubeconfig-overview.adoc:16]: I'm not sure what "the kubeconfig command tree" is. Do you mean the set of kubeconfig files that MicroShift creates for the user?

File: modules/microshift-kubeconfig-remote-con.adoc
Comment (dhellmann) [inline modules/microshift-kubeconfig-remote-con.adoc:9]: I was going to leave a comment here saying that we need to talk about "why" because without it this "what" information doesn't tell the user very much. The "why" comes later, so maybe this paragraph just needs to move down around what is now line 33?

File: modules/microshift-kubeconfig-remote-con.adoc
Comment (dhellmann) [inline modules/microshift-kubeconfig-remote-con.adoc:32]: We always generate the default kubeconfig files. The names in the `config.yaml` can be used to generate _more_ files.

**Review by dhellmann — COMMENTED** [review]
I'm worried that we're providing a lot of implementation detail about CAs and certificates here and as a result we are turning what should be very simple into something that seems very complicated. I've tried to suggest some specific changes inline, but generally we need to explain that

* MicroShift creates multiple client authentication files for the user automatically
* Each of those files works when accessing the API server using a specific hostname or IP address, so the user should pick the right file based on how they access the API server
* If the user needs additional authentication files using names other than the defaults, they can add those names to MicroShift's config.yaml to have it generate additional files

File: modules/microshift-kubeconfig-local-access.adoc
Comment (ShaunaDiaz) [reply modules/microshift-kubeconfig-local-access.adoc:13]: @dhellmann Would this be a good spot to call the other access modules we were discussing in https://issues.redhat.com/browse/OCPBUGS-10029?

File: _topic_maps/_topic_map_ms.yml
Comment (pacevedom) [reply _topic_maps/_topic_map_ms.yml:114]: For this to work you certainly need the config step to be complete. I think its a good fit to treat it as something an admin would do right after having everything properly configured and running.

File: microshift_networking/microshift-kubeconfig.adoc
Comment (pacevedom) [inline microshift_networking/microshift-kubeconfig.adoc:9]: Not sure about the `access type`. It is more or less implied in the IP address/hostname mentioned right after.

File: modules/microshift-kubeconfig-hosts-ips-config.adoc
Comment (pacevedom) [inline modules/microshift-kubeconfig-hosts-ips-config.adoc:19]: If configuring additional hostnames and IP addresses then it implies using external access, so there is a need to install `oc`.

File: modules/microshift-kubeconfig-hosts-ips-config.adoc
Comment (pacevedom) [inline modules/microshift-kubeconfig-hosts-ips-config.adoc:30]: Keep in mind that this path is the local kubeconfig, the remote ones follow a different path:
`/var/lib/microshift/resources/kubeadmin/<name/IP address>/kubeconfig`

The default remote access is the one under `/var/lib/microshift/resources/kubeadmin/$HOSTNAME/kubeconfig` which is always generated.

File: modules/microshift-kubeconfig-local-access.adoc
Comment (pacevedom) [reply modules/microshift-kubeconfig-local-access.adoc:9]: Yes, this file is only valid for localhost access now, it does not work if you copy it to a different host.
We can include that each access type (local, remote) is served a different certificate, each signed by a different CA and that is what the different kubeconfigs are for.

I would also remove the `local` in `allows access to the local API server`.

File: modules/microshift-kubeconfig-overview.adoc
Comment (pacevedom) [reply modules/microshift-kubeconfig-overview.adoc:16]: This might be a misunderstanding. Its the output of the `tree` command listing the contents of the kubeconfigs dir.

File: modules/microshift-kubeconfig-remote-con.adoc
Comment (pacevedom) [reply modules/microshift-kubeconfig-remote-con.adoc:32]: A kubeconfig using the hostname is always generated. Additional ones may be added via configuration. The one for the hostname is also editable to use the node IP instead.

File: modules/microshift-kubeconfig-hosts-ips-config.adoc
Comment (ShaunaDiaz) [reply modules/microshift-kubeconfig-hosts-ips-config.adoc:22]: Ah, OK. I thought the user could also change values in the remote access kubeconfig files also. I was wondering about that if they are autogenerated.

File: modules/microshift-kubeconfig-local-access.adoc
Comment (ShaunaDiaz) [reply modules/microshift-kubeconfig-local-access.adoc:13]: @pacevedom What do you think?

File: modules/microshift-kubeconfig-remote-con.adoc
Comment (ShaunaDiaz) [reply modules/microshift-kubeconfig-remote-con.adoc:32]: @pacevedom If the user edits that one, will the change be saved for the time MicroShift starts?

**Comment by ShaunaDiaz** [comment]
@dhellmann and @pacevedom Took another pass, PTAL. We can meet again if this PR just isn't quite there.

**Comment by ShaunaDiaz** [comment]
@dhellmann 1b7844c incorporates what we discussed; are we at LGTM?

File: modules/microshift-kubeconfig-generating-remote-kcfiles.adoc
Comment (dhellmann) [inline modules/microshift-kubeconfig-generating-remote-kcfiles.adoc:132]: If lines 67-69 are removed, we don't need this part either.

```suggestion
```

File: modules/microshift-kubeconfig-local-access.adoc
Comment (dhellmann) [inline modules/microshift-kubeconfig-local-access.adoc:9]: The path given is a file, not a directory.

```suggestion
The local access `kubeconfig` file is written to`/var/lib/microshift/resources/kubeadmin/kubeconfig`. This `kubeconfig` file provides access to the API server using `localhost`. Choose this file when you are connecting the cluster locally.
```

File: modules/microshift-kubeconfig-local-access.adoc
Comment (dhellmann) [inline modules/microshift-kubeconfig-local-access.adoc:20]: Trying to simplify and remove some implementation details to make the point a bit clearer.

```suggestion
The localhost kubeconfig file can only be used from a client connecting to the API server from the same host. The certificates in the file do not work for remote connections.
```

File: modules/microshift-kubeconfig-remote-con.adoc
Comment (dhellmann) [inline modules/microshift-kubeconfig-remote-con.adoc:24]:
```suggestion
Multiple remote access `kubeconfig` file values can be created for accessing the cluster via different IPs or hostnames. An additional `kubeconfig` file generates for each entry in the `apiServer.subjectAltNames` parameter. You can copy remote access `kubeconfig` files from the host during times of IP connectivity and then use them to access the API server from other workstations.
```

**Comment by ShaunaDiaz** [comment]
@dhellmann When we remove all of the content you reviewed as out for modules/microshift-kubeconfig-generating-remote-kcfiles.adoc, we end up with the same example file for what the user checked and what the user adds. This doesn't make sense. Please advise. See lines 29-40 and 53-64.

**Comment by dhensel-rh** [comment]
@ShaunaDiaz 
This documentation seems to duplicate similar wording found under **Installing from RPM**. I reviewed something similar in [58067](https://github.com/openshift/openshift-docs/pull/58067). Would it be best to link this documentation to the documentation under **Installing from RPM** instead of maintaining two areas of configuring ?

**Comment by ShaunaDiaz** [comment]
@dhensel-rh Yes, you are correct; the intent is to add those modules about accessing the cluster to this new content on how to use kubeconfig. I will rebase this PR to pick up the work we just did in https://github.com/openshift/openshift-docs/pull/58067. So you can focus on the kubeconfig portions of this PR.

**Comment by dhensel-rh** [comment]
Under **Accessing the MicroShift cluster remotely from a workstation** I think Step 5 can be removed ?

**Comment by dhellmann** [comment]
> Under **Accessing the MicroShift cluster remotely from a workstation** I think Step 5 can be removed ?

Step 5 can be removed, but step 4 also has to be changed to refer to the correct file, similar to how it works in the "Generating additional kubeconfig files for remote access" section.

**Review by dhellmann — COMMENTED** [review]
/lgtm

**Review by sheriff-rh — APPROVED** [review]
Really great job with this content, Shauna! No suggestions for improvement.

File: microshift_configuring/microshift-kubeconfig.adoc
Comment (ousleyp) [inline microshift_configuring/microshift-kubeconfig.adoc:5]: The `:context:` must be the same as the `id`, and ideally they should also match the file name for the assembly. In this case, all three could be `microshift-cluster-access-kubeconfig` because that matches the assembly title.

https://github.com/openshift/openshift-docs/blob/main/contributing_to_docs/doc_guidelines.adoc#assembly-file-metadata

File: microshift_configuring/microshift-kubeconfig.adoc
Comment (ousleyp) [inline microshift_configuring/microshift-kubeconfig.adoc:17]: The [modular docs assembly template](https://raw.githubusercontent.com/redhat-documentation/modular-docs/master/modular-docs-manual/files/TEMPLATE_ASSEMBLY_a-collection-of-modules.adoc) says that there should be a blank line after each `include::` statement. I know that not all of our assemblies follow that guidance, but you might want to fix it before merge.

File: modules/microshift-accessing-cluster-locally.adoc
Comment (ousleyp) [inline modules/microshift-accessing-cluster-locally.adoc:5]: Just noting that if you decide to rename the assembly file, you'll also want to update these "included in" comments to match the new name.

File: modules/microshift-kubeconfig-overview.adoc
Comment (ousleyp) [inline modules/microshift-kubeconfig-overview.adoc:17]: Formatting nit: the `'`s around config.yaml should be backticks.

File: modules/microshift-kubeconfig-generating-remote-kcfiles.adoc
Comment (ousleyp) [inline modules/microshift-kubeconfig-generating-remote-kcfiles.adoc:6]: Tiny nit, but IDs should be all lowercase for consistency.

File: modules/microshift-kubeconfig-generating-remote-kcfiles.adoc
Comment (ousleyp) [inline modules/microshift-kubeconfig-generating-remote-kcfiles.adoc:84]: Minor formatting issue: there is a space missing before "configuration values" that prevents the backticks around `apiServer.subjectAltNames` from rendering properly.

File: modules/microshift-kubeconfig-generating-remote-kcfiles.adoc
Comment (ousleyp) [inline modules/microshift-kubeconfig-generating-remote-kcfiles.adoc:16]: IIUC, we usually use the plural "Prerequisites" even when there is only one.

File: _topic_maps/_topic_map_ms.yml
Comment (ousleyp) [inline _topic_maps/_topic_map_ms.yml:120]: I realize this is commented out, but to be extra careful, I would remove anything extraneous, like lines 120 and 140.

File: modules/microshift-kubeconfig-generating-remote-kcfiles.adoc
Comment (ousleyp) [inline modules/microshift-kubeconfig-generating-remote-kcfiles.adoc:44]: Nit: this might not be causing any issues, but `[source, YAML]` should be `[source,yaml]` (applies to all instances)

File: modules/microshift-kubeconfig-overview.adoc
Comment (ousleyp) [inline modules/microshift-kubeconfig-overview.adoc:6]: The file name should match this title more closely

File: modules/microshift-kubeconfig-overview.adoc
Comment (ousleyp) [inline modules/microshift-kubeconfig-overview.adoc:21]: IIRC, we don't use `[source,bash]` for some reason. This should probably be `[source,terminal]` or `[source,text]` ([relevant section of the contrib guide](https://github.com/openshift/openshift-docs/blob/main/contributing_to_docs/doc_guidelines.adoc#code-blocks-command-syntax-and-example-output))

**Review by ousleyp — COMMENTED** [review]
A few things to fix, mostly minor. I also noted a couple of nits that wouldn't normally block merging, but they might be worth fixing along with the anchor IDs and such. Overall, the content looks good. 

Please reapply the merge-review-needed label when you're ready. Thanks!

File: modules/microshift-kubeconfig-generating-remote-kcfiles.adoc
Comment (ShaunaDiaz) [reply modules/microshift-kubeconfig-generating-remote-kcfiles.adoc:44]: Here, too, I have been told the opposite. I wonder if it doesn't matter?

File: modules/microshift-kubeconfig-overview.adoc
Comment (ShaunaDiaz) [reply modules/microshift-kubeconfig-overview.adoc:6]: Going to go with "= Kubeconfig files for configuring cluster access" since we have a prohibition on "overview" in titles.

**Pull request merged**
